### PR TITLE
pickle dictionary when it isn't JSON serializable

### PR DESCRIPTION
**Pull request by samhita-alla**
## Why are the changes needed?

This is useful for serializing dictionaries that aren't JSON serializable. To enable pickling the data, the `dict` type needs to be annotated with `allow_pickle` set to `True`:

```python
@task
def t1() -> Annotated[dict, kwtypes(allow_pickle=True)]
    ...
```

Pickling only happens if the dict isn't JSON serializable. 

## What changes were proposed in this pull request?

Modify `DictTransformer` to add support for pickling dictionaries and include relevant test cases. I've also updated the `BotoTask` output type to allow pickling when the dict isn't JSON serializable.

## Check all the applicable boxes 

- [ ] I updated the documentation accordingly.
- [ ] All new and existing tests passed.
- [x] All commits are signed-off.

### CODE REVIEW

File: flytekit/core/type_engine.py [hunk 1]
```diff
@@ -54,7 +55,14 @@
...
 )
-from flytekit.models.types import LiteralType, SimpleType, StructuredDatasetType, TypeStructure, UnionType
+from flytekit.models.types import (
+    LiteralType,
+    SimpleType,
+    StructuredDatasetType,
+    TypeStructure,
+    UnionType,
+)
+from flytekit.types.pickle import FlytePickle
```
Comment (dansola) [inline flytekit/core/type_engine.py:65]: I think this is a circular import. We might need to import it inside `dict_to_generic_literal()`

File: plugins/flytekit-aws-sagemaker/flytekitplugins/awssagemaker_inference/boto3_agent.py [hunk 2]
```diff
@@ -54,9 +57,20 @@ async def do(self, task_template: TaskTemplate, inputs: Optional[LiteralMap] = N
             inputs=inputs,
         )
 
-        outputs = None
+        outputs = {"result": {"result": None}}
```
Comment (pingsutw) [inline plugins/flytekit-aws-sagemaker/flytekitplugins/awssagemaker_inference/boto3_agent.py:60]: Isn't None Json serializable? could we create a dummy class here

Screenshot 2024-05-08 at 7 12 33 PM

File: flytekit/core/type_engine.py [hunk 3]
```diff
@@ -1679,22 +1677,53 @@ def get_dict_types(t: Optional[Type[dict]]) -> typing.Tuple[Optional[type], Opti
...
     def get_literal_type(self, t: Type[dict]) -> LiteralType:
         """
         Transforms a native python dictionary to a flyte-specific ``LiteralType``
         """
-        tp = self.get_dict_types(t)
+        if get_origin(t) is Annotated:
+            base_type, *_ = self.extract_types_or_metadata(t)
+            tp = get_args(base_type)
+        else:
+            tp = self.extract_types_or_metadata(t)
```
Comment (pingsutw) [inline flytekit/core/type_engine.py:1725]: nit: could we move the logic here into `self.extract_types_or_metadata`? something like

```python
def extract_types_or_metadata(self, python_type):
    if get_origin(python_type) is Annotated:
        python_type = get_args(python_type)[0]
    ... # original code here
```

so we can change the code here to
```python
tp = self.extract_types_or_metadata(t)
```

File: flytekit/core/type_engine.py [hunk 4]
```diff
@@ -1710,21 +1739,37 @@ def to_literal(
...
         return Literal(map=LiteralMap(literals=lit_map))
 
     def to_python_value(self, ctx: FlyteContext, lv: Literal, expected_python_type: Type[dict]) -> dict:
         if lv and lv.map and lv.map.literals is not None:
-            tp = self.get_dict_types(expected_python_type)
+            if get_origin(expected_python_type) is Annotated:
+                base_type, *_ = self.extract_types_or_metadata(expected_python_type)
+                tp = get_args(base_type)
+            else:
+                tp = self.extract_types_or_metadata(expected_python_type)
```
Comment (pingsutw) [inline flytekit/core/type_engine.py:1771]: ditto

File: plugins/flytekit-aws-sagemaker/flytekitplugins/awssagemaker_inference/boto3_agent.py (hunk 2 above)
Comment (samhita-alla) [reply plugins/flytekit-aws-sagemaker/flytekitplugins/awssagemaker_inference/boto3_agent.py:60]: modified it to `None`

File: flytekit/core/type_engine.py (hunk 3 above)
Comment (samhita-alla) [reply flytekit/core/type_engine.py:1725]: i'm not sure if this is possible 'cause i don't want to always return the dict key and value types. i also need access to the annotation metadata to check if `allow_pickle` is set to `True`.

```python
def is_pickle(self, python_type: Type[dict]) -> typing.Tuple[bool, Type]:
        base_type, *metadata = self.extract_types_or_metadata(python_type)

        for each_metadata in metadata:
            if isinstance(each_metadata, OrderedDict):
                allow_pickle = each_metadata.get("allow_pickle", False)
                return allow_pickle, base_type

        return False, base_type
```

Comment (samhita-alla) [reply flytekit/core/type_engine.py:1725]: so if `extract_types_or_metadata` returns the args, i can either fetch the metadata or the actual dict types depending on what i need. makes sense?

Comment (samhita-alla) [reply flytekit/core/type_engine.py:1725]: i could probably encapsulate this logic in a separate function to avoid repetition. ok?

**Review by pingsutw — APPROVED** [review]

**Pull request merged**